src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -2895,6 +2895,9 @@ else if (!skipNonDiamondPath) {
 
             if (tree.constructor != null && tree.constructor.kind == MTH)
                 owntype = clazztype;
+                if (owntype.getMetadata(TypeMetadata.NullMarker.class) == null) {
+                    owntype = owntype.addMetadata(new TypeMetadata.NullMarker(NullMarker.NOT_NULL)); // constructor invocations are always null restricted
```

Comment:
I note an asymmetry here: for constructor calls, we annotate the type to say it's non-null. But for constant types we don't, and we instead look as to whether there's constant metadata. Perhaps an alternative to the suggestion I gave above is to simply set the null-restricted metadata in the visitLiteral method, so that everything is explicit.